### PR TITLE
Use mTLS mode enum to decide authentication result.

### PR DESCRIPTION
**Pull request by diemtvu**
We want to support different authentication mode in order to support mTLS migrations.

Issue: https://github.com/istio/istio/issues/5059

### CODE REVIEW

File: src/envoy/http/authn/authenticator_base.cc [hunk 1]
```diff
@@ -34,16 +34,33 @@ AuthenticatorBase::~AuthenticatorBase() {}
...
+  // - STRICT: must be TLS with valid certificate.
+  switch (mtls.mode()) {
+    case iaapi::MutualTls::PERMISSIVE:
+      return true;
+    case iaapi::MutualTls::TLS_PERMISSIVE:
+      return connection->ssl() != nullptr;
+    case iaapi::MutualTls::STRICT:
+      return has_user;
+    default:
+      ENVOY_LOG(warn, "Invalid mTLS mode {}",
```
Comment (qiwzhang) [inline src/envoy/http/authn/authenticator_base.cc]: use NOT_REACHED 

defined in common/common/assert.h

Comment (diemtvu) [reply src/envoy/http/authn/authenticator_base.cc]: Done

**Pull request merged**